flink-connectors/flink-connector-upsert-test/src/test/java/org/apache/flink/connector/upserttest/sink/UpsertTestSinkWriterITCase.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.upserttest.sink;
+
+import org.apache.flink.api.common.serialization.AbstractDeserializationSchema;
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+import org.apache.flink.api.common.serialization.SerializationSchema;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.util.TestLoggerExtension;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.io.TempDir;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for {@link UpsertTestSinkWriter}. */
+@ExtendWith(TestLoggerExtension.class)
+class UpsertTestSinkWriterITCase {
+
+    @Test
+    public void testWrite(@TempDir File tempDir) throws Exception {
+        File outputFile = new File(tempDir, "records.out");
+        SerializationSchema<Tuple2<String, String>> keySerializationSchema =
+                element -> element.f0.getBytes();
+        SerializationSchema<Tuple2<String, String>> valueSerializationSchema =
+                element -> element.f1.getBytes();
+
+        try (final UpsertTestSinkWriter<Tuple2<String, String>> writer =
+                new UpsertTestSinkWriter<>(
+                        outputFile, keySerializationSchema, valueSerializationSchema)) {
+            for (int i = 0; i < 10; i++) {
+                writer.write(Tuple2.of("Key #" + i, "Value #" + i), null);
+            }
+        }
+
+        DeserializationSchema<String> keyDeserializationSchema = new StringDeserializationSchema();
+        DeserializationSchema<String> valueDeserializationSchema =
+                new StringDeserializationSchema();
+
+        Map<String, String> resultMap =
+                BinaryFileUtil.readRecords(
+                        outputFile, keyDeserializationSchema, valueDeserializationSchema);
+
+        for (int i = 0; i < 10; i++) {
+            assertThat(resultMap).containsEntry("Key #" + i, "Value #" + i);
+        }
+    }
+
+    @Test
+    public void testWriteOnCheckpoint(@TempDir File tempDir) throws Exception {
```

Comment:
Can you extract the common code for both tests? In the end, the setup and assertion are equal the only difference is in one scenario you verify that on closing all records are flushed and in the other, you are verifying the behavior on flush. It might also make sense to only have one parameterized test based on the operation of the `UpsertTestSinkWriter`.